Add "prepare" and "check" watchers

mergeconflict opened the pull request:
Adds two new callbacks: "prepare" watchers, which fire immediately before we poll for I/O, and "check" watchers, which fire immediately after we finish polling and before we process events. This allows other event loops to be embedded into libevent's, and enables certain performance monitoring.

Closes: #710

mergeconflict commented on CMakeLists.txt:
```diff
@@ -737,6 +737,7 @@ set(HDR_PUBLIC
     include/event2/event.h
     include/event2/event_compat.h
     include/event2/event_struct.h
+    include/event2/event_watcher.h
```
I think there are non-CMake builds I need to support as well... can somebody help direct me to the other build files I need to update? Thanks!

azat replied:
autotools (which is primary build system for libevent right now) -- `Makefile.am`

*P.S. I do not ask you to fix manually written files for nmake, since I'm going to remove them*

mergeconflict commented on evwatcher.c:
```diff
@@ -0,0 +1,70 @@
+event_watcher_free(struct event_watcher *watcher)
+{
+	TAILQ_REMOVE(&watcher->base->event_watchers[watcher->type], watcher, next);
+	free(watcher);
```
Oops, I think this should be `mm_free`.

azat replied:
yep

mergeconflict commented on include/event2/event_watcher.h:
```diff
@@ -0,0 +1,146 @@
+		to retain it outside the scope of the callback (e.g. to measure the duration between callbacks),
+		make a copy.
+	 */
+	struct timeval now;
```
Question: is it ok to include this here, or do we prefer users to use `event_base_update_cache_time` and `event_base_gettimeofday_cached`?

azat replied:
I would prefer to drop this from this structure

azat replied:
BTW why it has been added here? This prepare/check thing does not looks like it is care about cached time, or maybe I'm missing something?

mergeconflict replied:
It was mainly for convenience, not strictly necessary. I included it because one of the metrics we'd like to track is the duration between a `check` callback and the following `prepare` callback (on the next iteration of the event loop). But I know we can get that another way. The only really necessary bit is the `timeout` below, which there doesn't seem to be any other way to get.

mergeconflict commented on include/event2/event_watcher.h:
```diff
@@ -0,0 +1,146 @@
+		value -- essentially the expected polling duration -- against the actual polling duration
+		measured by comparing the "now" times reported in prepare and check.
+	 */
+	const struct timeval *timeout;
```
Note: this is the key bit of information we need to expose in this PR.

mergeconflict commented on include/event2/event_watcher.h:
```diff
@@ -0,0 +1,146 @@
+
+/** @file event2/event_watcher.h
+
+	"Prepare" and "check" watchers. A "prepare" watcher is a callback that fires immediately before
```
If this PR is merged, how do we go about updating the user docs (e.g. http://www.wangafu.net/~nickm/libevent-book)?

azat replied:
That will be great! Will you? (and also one good example will be great, like duration histograms that was mentioned in #710)

azat replied:
BTW the book is [here](https://github.com/nmathewson/libevent-book)

azat replied:
(but let's deal with the patches first)

mergeconflict replied:
Sounds good, thanks!

mergeconflict commented:
cc @htuch

mergeconflict commented on evwatcher.c:
```diff
@@ -0,0 +1,70 @@
+struct event_base *
+event_watcher_base(struct event_watcher *watcher)
+{
+	return watcher->base;
```
Needs test coverage.

mergeconflict commented on evwatcher.c:
```diff
@@ -0,0 +1,70 @@
+{
+	struct event_watcher *watcher = mm_malloc(sizeof(struct event_watcher));
+	if (!watcher) {
+		return NULL;
```
Coveralls notes this line lacks test coverage... Do we have any way of mocking mm_malloc to test the behavior when allocation fails?

azat replied:
Sure, `event_set_mem_functions()`

azat commented on .gitignore:
```diff
@@ -5,6 +5,7 @@
 .#*
 *~
 *.swp
+.vscode
```
I would prefer if this line will be in a separate patch (but keep it in this merge)
The main reason this is to keep main patch contains only prepare/check bits

azat commented on event-internal.h:
```diff
@@ -205,6 +206,31 @@ struct event_once {
 
+#define EVENT_WATCHER_PREPARE_TYPE 0
+#define EVENT_WATCHER_CHECK_TYPE   1
+#define EVENT_WATCHER_TYPES        2
```
Can we name it something like `EVENT_WATCHES_TYPE_MAX` or something similar, it makes meaning of this variable cleaner (that this is just the last of the watches defines)

azat replied:
also can you come up with more compact name, for example how about `EVW_PREPARE` (but this is just an advice and maybe it is too compact)

azat commented on event.c:
```diff
@@ -939,6 +946,15 @@ event_base_free_(struct event_base *base, int run_finalizers)
+		while (!TAILQ_EMPTY(&base->event_watchers[i])) {
+			watcher = TAILQ_FIRST(&base->event_watchers[i]);
+			TAILQ_REMOVE(&base->event_watchers[i], watcher, next);
+			free(watcher);
```
`mm_free` here too

azat commented on evwatcher.c:
```diff
@@ -0,0 +1,70 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
I think that this header can be ommited... or just remove first two copyrights

mergeconflict replied:
Ok, I'll take off the two copyrights from the three new files... Let me know if I misunderstood.

azat commented on include/event2/event_watcher.h:
```diff
@@ -0,0 +1,146 @@
+struct event_watcher;
+
+/**
+	Contextual information passed from event_base_loop to the "prepare" watcher callbacks. We define
```
Some of your lines are pretty long (not only here), can you make it the same style as include/event2/event.h  (although we do not have any strict restrictions on line length, unless it is *sane*)

mergeconflict replied:
Sure, no problem. Out of curiosity, what tab-width do you use in your editor? I was using a tab-width of 2, but now I'm guessing that's much smaller than yours :)

azat replied:
Mine is four, but I guess that lots of code had been written with 6

azat commented on include/event2/event_watcher.h:
```diff
@@ -0,0 +1,146 @@
+	this as a struct rather than individual parameters to the callback function for the sake of future
+	extensibility.
+ */
+struct prepare_watcher_cb_info {
```
- this does not contains `event_` prefix (and not only this one)
- but please do not call it `event_prepare_watcher_cb_info`, tooooo long
- and if this `struct` exposed then you will not have ability to change it, maybe it is better to hide it in module part and provide getter for the only field that it has, in this case user will not allow to change our timeout (since it has a pointer to our internal variable) which is not good from some point of vew

azat commented on test/test-prepare-check.c:
```diff
@@ -0,0 +1,153 @@
+static struct timeval start_time = { 0, 0 };
+static struct timeval end_time = { 0, 0 };
+
+#define log_check_callback \
```
can you wrap this to `do { ... } while (0);` or at least `{}` block? (like in other places)

azat commented on test/test-prepare-check.c:
```diff
@@ -0,0 +1,153 @@
+	/* this will free the watchers that weren't already freed in timeout_callback */
+	event_base_free(base);
+	return (0);
+}
```
lines are also pretty long

azat commented on test/test-prepare-check.c:
```diff
@@ -0,0 +1,153 @@
+
+	/* this will free the watchers that weren't already freed in timeout_callback */
+	event_base_free(base);
+	return (0);
```
please do not wrap the value (personally I think that this style came from some parts that had been migrated from other places, and that was a long time ago)

azat replied:
so just `return 0;`

mergeconflict commented:
@azat I still haven't added the example under `sample/` you asked for -- I'll work on that next...

mergeconflict commented on event.c:
```diff
@@ -1987,6 +2003,14 @@ event_base_loop(struct event_base *base, int flags)
 		event_queue_make_later_events_active(base);
 
+		/* Invoke prepare watchers before polling for events */
+		EVBASE_RELEASE_LOCK(base, th_base_lock);
```
Another question: when my test was a standalone executable, it seemed there was no problem calling `event_base_gettimeofday_cached()` from within a watcher callback. But when I converted it into `regress` testcases, it broke: the lock was already held in the call, and it failed:

```
[err] /usr/local/google/home/mergeconflict/github/libevent/evthread.c:239: Assertion lock->count == 1 failed in evthread_debug_lock_mark_locked

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt    
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff66453fa in __GI_abort () at abort.c:89
#2  0x00007ffff69e49bf in event_exit (errcode=-559030611) at /usr/local/google/home/mergeconflict/github/libevent/log.c:96
#3  0x00007ffff69e4fb4 in event_errx (eval=-559030611, fmt=0x7ffff69e8774 "%s:%d: Assertion %s failed in %s")
    at /usr/local/google/home/mergeconflict/github/libevent/log.c:153
#4  0x00007ffff69dd5e4 in evthread_debug_lock_mark_locked (mode=0, lock=0x715560)
    at /usr/local/google/home/mergeconflict/github/libevent/evthread.c:239
#5  0x00007ffff69dcea4 in debug_lock_lock (mode=0, lock_=0x715560) at /usr/local/google/home/mergeconflict/github/libevent/evthread.c:261
#6  0x00007ffff69cd683 in event_base_gettimeofday_cached (base=0x7150e0, tv=0x7fffffffda30)
    at /usr/local/google/home/mergeconflict/github/libevent/event.c:445
```

Releasing the lock before invoking the watchers fixes the issue, and it seemed like the right thing to do since we also release it when invoking user callbacks for active events, but I want to double-check with you that this makes sense.

mergeconflict replied:
I also wonder why it worked at all when my test was a standalone executable...

azat replied:
> when my test was a standalone executable, it seemed there was no problem calling event_base_gettimeofday_cached() from within a watcher callback

libevent do not thread safe by default (since this is overhead), you can enable it with `evthread_use_pthreads()` this is the first part, see below for the second

> But when I converted it into regress testcases, it broke: the lock was already held in the call, and it failed:

Second part is that `regress` enables lock debugging (`evthread_enable_lock_debugging()`) and that is why this assert pops up (you can disable it by installing `EVENT_NO_DEBUG_LOCKS` environment variable) instead of dead locking the program (since relocking of the non-recursive mutex is most of the time this is deadlock, but strictly speaking it is "undefined behavior")

>Releasing the lock before invoking the watchers fixes the issue, and it seemed like the right thing to do since we also release it when invoking user callbacks for active events, but I want to double-check with you that this makes sense

Indeed, my mistake that I didn't noticed this during looking into this yesterday (actually I didn't look for the correctness that time, just "conceptually")
Any callback that is called from the event loop should not hold the base mutex.

mergeconflict replied:
Thinking a bit about overhead... I'm wondering if we should gate this new feature behind a compilation flag, like:
```C
#ifdef EVENT__ENABLE_WATCHERS
...
#endif
```
so that you don't pay any performance cost (especially around locking/unlocking) if you're not going to use the feature. What do you think?

azat commented on include/event2/event_watcher.h:
```diff
@@ -29,98 +26,63 @@
  */
 EVENT2_EXPORT_SYMBOL
-struct event_watcher *event_watcher_prepare_new(struct event_base *base, prepare_watcher_cb callback);
+struct event_watcher *event_watcher_prepare_new(struct event_base *base, event_prepare_cb callback);
```
looking into this again, and priorities support for this feature came to my mind, i.e. if we want this be generic, since if some library has event_base and registered watches and user want to call something before that library watches he cannot do this right now.

although this is pretty specific thing, and to avoid delaying merging this, we can at least reserve argument in the API for this (to make further changes ABI compatible) and return an error if `prio != 0`

mergeconflict replied:
Let me suggest an alternative approach: if we want to support priority for watchers later, we could add a function `event_watcher_priority_set`, similar to the existing `event_priority_set`. That would keep the `event_watcher_*_new` APIs minimal, and allow you to change the priority of a watcher...

azat replied:
Yep, sounds great

mergeconflict commented on test/regress_watcher.c:
```diff
@@ -0,0 +1,231 @@
+static void
+timeout_callback(evutil_socket_t fd, short events, void *arg)
+{
+	printf("timeout_callback\n");
```
I actually think it's best for me to just remove these - they don't really indicate success or failure, they were just helpful debugging info for me while I was learning how the event loop works.

mergeconflict commented on event-internal.h:
```diff
@@ -206,16 +206,38 @@ struct event_once {
+		The timeout duration passed to the underlying implementation's `dispatch`.
+		See event_prepare_get_timeout.
+	 */
+	const struct timeval *timeout;
```
Hmmm. I don't feel strongly about this, but let me argue a little bit: I was thinking that the `const` provides enough safety -- if somebody casts away the `const`, it's their own fault. And maybe it's not actually a bad thing for users to modify the timeout? I don't know :)

Anyway, if you disagree, then I'm certainly ok with copying the timeout in the getter. The only trick is that the timeout may be `NULL` (e.g. if we're not blocking or we have no timeout events), so I'd need the getter to look something like:

```
int
event_prepare_get_timeout(const struct event_prepare_cb_info *info, struct timeval *timeout) {
	if (info->timeout) {
		*timeout = *(info->timeout);
		return 1;
	}
	return 0;
}
```

Does that look right?

mergeconflict commented on test/regress_watcher.c:
```diff
@@ -0,0 +1,231 @@
+	    without a recursive lock.
+	  - If this test is compiled with ASAN or similar, this test also
+	    illustrates that event_base_free will free any watchers not
+	    previously freed by event_watcher_free.
```
I took out the extra two spaces but left the `-` bullets... let me know if that looks ok to you.

azat replied:
I'm talking about that "common" comments style is one of the following:

- 4 spaces for alignment

```c
/** Brief description.

    foo bar

    @return 0 on success, -1 on failure
 */
```

- uses `*` at the beginning of the line
```c
/** Brief description
 *
 * Foo bar
 *
 * @return 0 on success, -1 on failure
 */
```

- but what I hadn't seen is using tabs for multi-line comments

azat commented:
> Thinking a bit about overhead... I'm wondering if we should gate this new feature behind a compilation flag, like:

The overhead should be neglible (one more lock/unlock, BTW we can get
rid off the impact by checking if there are any watchers and if not we
can skip lock/foreack/unlock, but again I don't think that even this
simple condition worth it) since event_base_loop() should not be called
tons of times (by those who cares about performance).

Also introducing yet another compilation option increases coverage
matrix (and it is already huge enough).

azat commented:
> Hmmm. I don't feel strongly about this, but let me argue a little bit: I was thinking that the `const` provides enough safety -- if somebody casts away the `const`, it's their own fault. And maybe it's not actually a bad thing for users to modify the timeout? I don't know :)

I'm not sure that we should allow changing this time, and even if we
should, then explicitly via some API, like
event_prepare_set_{loop/base}_timeout() not via hackish casts
(personally I, don't care about architectures other then x86/amd64, but
this const-cast-away can break something for them).

But maybe I'm too suspicious about this...

> int
> event_prepare_get_timeout(const struct event_prepare_cb_info *info, struct timeval *timeout) {
> 	if (info->timeout) {
> 		*timeout = *(info->timeout);
> 		return 1;
> 	}
> 	return 0;
> }

Looks good.

azat commented on event.c:
```diff
@@ -737,6 +735,11 @@ event_base_new_with_config(const struct event_config *cfg)
 #endif
 
+	/* initialize event watcher lists */
+	for (i = 0; i < EV_WATCHER_MAX; ++i) {
```
style: no `{}` for oneline body

i.e.

- correct
```c
for (;;)
    bar();
```

- not common:
```c
for (;;) {
    bar();
}
```

mergeconflict commented:
I'm still working on the timing histogram example. Should I just add that in a separate pull request?

mergeconflict commented:
Ok... I've added an example of how we might use prepare and check watchers to collect timing statistics. There's a fun implementation of constant-space histograms in there... it might be nice to turn that into a proper library at some point. @htuch, @mattklein123 - maybe you'd be interested in having a look at the example as well.

I haven't finished addressing all the style comments yet. I'll make another pass tomorrow.

mergeconflict commented:
Hi @azat, thanks for all your feedback! I think I've addressed all the style issues, I've added a "what's new" entry, and I've updated the issue description.

I just want to double-check -- you want me to `git rebase -i` locally to squash all the commits, using this issue description as the commit message, and then force-push the branch -- is that right?

azat commented:
> I think I've addressed all the style issues, I've added a "what's new" entry, and I've updated the issue description.

Thanks! Will take a closer look today or at least tomorrow

> I just want to double-check -- you want me to git rebase -i locally to squash all the commits, using this issue description as the commit message, and then force-push the branch -- is that right?

Yep git rebase locally, squash all the commits, update commit
description and git push --force

azat commented on evwatcher.c:
```diff
@@ -0,0 +1,76 @@
+	watcher->base = base;
+	watcher->type = type;
+	watcher->callback = callback;
+	TAILQ_INSERT_TAIL(&base->watchers[type], watcher, next);
```
missing base lock/unlock

azat commented on evwatcher.c:
```diff
@@ -0,0 +1,76 @@
+void
+event_watcher_free(struct event_watcher *watcher)
+{
+	TAILQ_REMOVE(&watcher->base->watchers[watcher->type], watcher, next);
```
missing base lock/unlock

azat commented on sample/watcher-timing.c:
```diff
@@ -0,0 +1,315 @@
+eq(double a, double b)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wfloat-equal"
```
can you show that this is required?

mergeconflict replied:
I get the following warning without it:
```
sample/watcher-timing.c:43:11: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
        return a == b;
               ~ ^  ~
```

azat commented on sample/watcher-timing.c:
```diff
@@ -0,0 +1,315 @@
+histogram_query(const struct histogram *h, double quantile)
+{
+	unsigned lhs = 0, rhs = 0;
+	struct bin lhs_bin, rhs_bin;
```
can you initialize them, to avoid `[-Wmaybe-uninitialized]` ? (you can see this warnings with release build type)

azat replied:
hm, and looks like this is not false-positive

azat commented on evwatcher.c:
```diff
@@ -0,0 +1,76 @@
+}
+
+int
+event_prepare_get_timeout(const struct event_prepare_cb_info *info, struct timeval *timeout)
```
hm, looking into this again, and `event_prepare_get_timeout`does not looks like a good naming, how about `evwatch_prepare_get_timeout()` ? (and convert the whole patch into this)

mergeconflict replied:
Do you mean to change all:
```C
struct event_watcher => struct evwatch
struct event_prepare_cb_info => struct evwatch_prepare_cb_info
struct event_check_cb_info => struct evwatch_check_cb_info

event_prepare_cb => evwatch_prepare_cb
event_check_cb => evwatch_check_cb

event_watcher_prepare_new => evwatch_prepare_new
event_watcher_check_new => evwatch_check_new
event_watcher_base => evwatch_base
event_watcher_free => evwatch_free
```
Should I also rename `event_watcher.h` to `evwatch.h`? I don't understand the naming conventions here... Or do you just want me to rename the one function?

azat replied:
Yep, all of them, i.e. like `evbuffer` was done
And if we are doing this like `evbuffer`, then the file should be called `<event2/watch.h>`

azat commented on sample/watcher-timing.c:
```diff
@@ -0,0 +1,315 @@
+	event_watcher_prepare_new(base, &on_prepare);
+	event_watcher_check_new(base, &on_check);
+	event_base_dispatch(base);
+
```
all below (`*free()`) will never be reached, but what you can do is add a handler for `SIGINT` (see `evsignal_*()`) and call `event_base_loopexit()` on it, then you will got no leaks on termination by interrupting.

azat commented:
> +#pragma GCC diagnostic ignored "-Wfloat-equal"
>
> I get the following warning without it:
>
> sample/watcher-timing.c:43:11: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
>         return a == b;
>                ~ ^  ~

Got it, thanks! (yet another difference between autotools and cmake,
will fix this later).

mergeconflict commented:
@azat updated and rebased. The lock/unlock was a good catch. Thanks again for the thorough review!

mergeconflict commented:
Hi @azat, everything's updated. Let me know if there's anything else!

azat commented:
@mergeconflict everything looks good!

And while thinking about this one more time, I think that user-passed argument will not be superfluous, i.e. `evwatch_prepare_new(struct event_base *base, evwatch_prepare_cb callback, void *arg)`, can you add this? (and I think that this is the last bit, sorry for not noticing this before)

mergeconflict commented:
> And while thinking about this one more time, I think that user-passed argument will not be superfluous, i.e. `evwatch_prepare_new(struct event_base *base, evwatch_prepare_cb callback, void *arg)`, can you add this? (and I think that this is the last bit, sorry for not noticing this before)

Sure, no problem. I was also actually thinking it might be handy. I'll put the `*arg` as an additional argument in the callbacks, after the `evwatch_*_cb_info` to stay consistent with how the `evbuffer` callbacks are designed.

azat commented:
So everything looks great, will apply this once the CI will be passed (or almost passed)
Thanks for the quick reaction!

azat commented:
One more thing pops up `evwatch_check_cb_info` is empty, and while this is accepted by gcc it is not strictly valid and VS do not pass this:
```
c:\projects\libevent\event-internal.h(221): error C2016: C requires that a struct or union has at least one member [C:\projects\libevent\build-cmake\event_core_shared.vcxproj]
```

Maybe this can be suppressed with some kind of pragma, but I think that better option will be to add one field there instead, like the time the dispatch took

azat commented:
And commit message should be formatted appropriate, i.e.:

```
subject
<   blank line   >
body
```

since otherwise git will think that the whole commit message is the subject (without empty line)

(first link from google https://chris.beams.io/posts/git-commit/#seven-rules)

mergeconflict commented:
> Maybe this can be suppressed with some kind of pragma, but I think that better option will be to add one field there instead, like the time the dispatch took

Interesting, that's actually very helpful for us -- it would mean that we don't have to save the prepare and check timestamps externally. I'm thinking to rename the prepare timeout `expected_timeout` and call the check timeout `actual_timeout`, so that it's obvious to users that you can compare the actual and expected values. What do you think?

azat commented:
>Interesting, that's actually very helpful for us -- it would mean that we don't have to save the prepare and check timestamps externally. I'm thinking to rename the prepare timeout expected_timeout and call the check timeout actual_timeout, so that it's obvious to users that you can compare the actual and expected values. What do you think?

I like the idea in general, but don't like such long fields. I guess it is fine here

mergeconflict commented:
Ok... I've implemented timing, but I haven't squashed the commits yet, since I'm not sure if you'll be ok with it. Specifically, to get accurate timing, I've had to add two `evutil_gettimeofday` calls before and after `evsel->dispatch`, which might be unacceptable overhead in some cases. I tried using the cached time, but it was too inaccurate (like 3000us compared to about 1000us expected).

Of course I'd be very happy to keep this implementation, since it supports my use case directly, but I understand if it's not good for others. If that's the case, then we'll solve the Win32 empty struct issue by simply adding a dummy unused field instead.

mergeconflict commented:
Hmm... I just noticed I have a link error on Windows:
```
LINK : fatal error LNK1104: cannot open file 'm.lib' [C:\projects\libevent\build-cmake\watch-timing.vcxproj]
```
Trying a fix...

azat commented:
>to get accurate timing, I've had to add two evutil_gettimeofday calls before and after evsel->dispatch, which might be unacceptable overhead in some cases

Well if we are talking about modern linux (but maybe I missed something, since this is 5min experience) then I don't think that cached version is worth it, and just out of curiosity I wrote [simple test](https://gist.github.com/azat/2168ee8cc3a29397be63dcf9894f7075#file-le-gettimeofday-c) and this assumption is correct (actually non cached version is a little bit faster)

But we have lots of other supported architectures, so I let's use cached version here and this will be addressed later

(because since we do have this `event_base_gettimeofday_cached()` now we should use it, if we do not want it, it should be removed completely in one pass)

mergeconflict commented:
> But we have lots of other supported architectures, so I let's use cached version here and this will be addressed later

The issue with using the cached time is that it's too inaccurate, at least for the main use case we care about in Envoy. The goal of measuring the actual vs. expected polling duration is that if you compare the two, the actual duration should be *very close* to the expected duration - within one kernel timeslice usually. If the delay is larger, it's an indication that your kernel is swamped. Using the cached time will give the misleading impression that something's wrong...

My suggestion is, let's leave this out for now. In the future, if we get rid of the cached time (since your experimental results seem good), then we can put the accurate timing here. Does that sound ok?

azat commented:
> My suggestion is, let's leave this out for now. In the future, if we get rid of the cached time (since your experimental results seem good), then we can put the accurate timing here. Does that sound ok?

Agree

mergeconflict commented:
Ok, cool. I've updated:

* Revert the timing thing
* Fix the empty struct issue by adding a dummy unused field
* Fix the git commit message
* Squashed and rebased against the latest changes in master

azat commented:
Applied, thanks for the effort!

mergeconflict commented:
Thanks @azat!
